### PR TITLE
[CSApply] Fix an issue with the @objc selector fix-it incorrectly being applied

### PR DESCRIPTION
**Pull request by theblixguy**
This PR fixes an issue where the `@objc` fix-it would be applied to the method declaration in the protocol. This would happen if you provided a default implementation of the method via a protocol extension and tried using it as a selector inside another method within the same extension. The fix-it would then be applied to the method on the protocol.

```
protocol Foo {
  func bar()
  func baz() // The `@objc` fix-it would then be applied here (incorrect).
}

extension Foo {
  func bar() {

   // This would trigger an error diagnostic that would offer the fix-it.
    let sel = #selector(baz)
  }

  func baz() {
    print("Hello, world)")
  }
}
```

With this fix, no fix-it will be applied as it might not actually work and might not be what the user expects.

```
protocol Foo {
  func bar()
  func baz()
}

extension Foo {
  func bar() {

    // This would trigger an error diagnostic. No fix-it will be offered.
    let sel = #selector(baz)
  }

  func baz() {
    print("Hello, world)")
  }
}
```

Resolves [SR-9391](https://bugs.swift.org/browse/SR-9391).

### CODE REVIEW

**Comment by theblixguy** [comment]
cc @harlanhaskins @jrose-apple @slavapestov @jckarter

**Comment by CodaFi** [comment]
Could you please add a test.

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -4196,12 +4196,26 @@ namespace {
...
+	// we can insert the @objc fix-it on the protocol declaration, instead
+	// of on the method declaration (as that's not allowed). This fixes an
+	// issue where you provide a default implementation of a method via a
+	// protocol extension and the fix-it ends up being on the method
+	// declaration on the protocol.
+				
+	// FIXME: Put the fix-it on the method declaration in the protocol
+	// extension, instead of the protocol itself?
+	auto methodContextDecl = foundDecl->getDeclContext()->getAsDecl();
+	bool isMethodDeclaredInProtocol = methodContextDecl->getKind() == DeclKind::Protocol;
```
Comment (CodaFi) [inline lib/Sema/CSApply.cpp]: You can use `foundDecl->getDeclContext()->getExtendedProtocolDecl()` here.

**Comment by CodaFi** [comment]
There's a really narrow case where this still won't be the right fixit 

```
protocol Foo {
  associatedtype Uhoh
  func bar()
  func baz()
}
```

If you try to mark `Foo` `@objc` you're in the same boat.

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Actually, `foundDecl->getDeclContext()->getExtendedProtocolDecl()` returns null, so I think I need to use `getSelfProtocolDecl()` instead.

**Comment by jrose-apple** [comment]
I don't think we want to suggest making a Swift protocol `@objc` when it wasn't before. That's a drastic change in many cases, and it doesn't solve the default implementation problem. Just not offering the fix-it is probably good enough, and if you wanted to go further you could add a note like "members of protocol extensions cannot be exposed to Objective-C".

**Comment by theblixguy** [comment]
@jrose-apple Ah ok, sounds good. I’ll make the change. I was wondering if we should offer an option to add `@objc` to the user, in case that’s what they want?

**Comment by jrose-apple** [comment]
Oh, right, this is for `#selector`, not trying to satisfy another protocol. Sorry! I guess the fix-it does make sense but it's worth checking that the decl context is actually a ProtocolDecl, not just that it has a SelfProtocolDecl (because that covers extensions too).

File: lib/Sema/CSApply.cpp [hunk 2]
```diff
@@ -4196,14 +4196,22 @@ namespace {
...
-                    foundDecl->getDescriptiveKind(), foundDecl->getFullName())
+          // If the method declaration lies in a protocol and we're providing
+          // a default implementation of the method through a protocol extension
+          // then insert the fixit on protocol, rather than on the method in the
+          // protocol declaration (not allowed).
+          auto protocolDecl = foundDecl->getDeclContext()->getSelfProtocolDecl();
+          
+          tc.diagnose(E->getLoc(), diag::expr_selector_not_objc,
+                      foundDecl->getDescriptiveKind(), foundDecl->getFullName())
           .highlight(subExpr->getSourceRange());
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Nitpick: these chaining lines are supposed to be indented. (In our usual style continuation lines are double-indented, but we get sloppy about it for diagnostics sometimes because we chain so frequently, and in this case the dots are lining up.)

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Done! 👍

File: lib/Sema/CSApply.cpp [hunk 3]
```diff
@@ -4198,16 +4198,16 @@ namespace {
       if (!method->isObjC()) {
           // If the method declaration lies in a protocol and we're providing
           // a default implementation of the method through a protocol extension
-          // then insert the fixit on protocol, rather than on the method in the
+          // then insert the fix-it on protocol, rather than on the method in the
           // protocol declaration (not allowed).
-          auto protocolDecl = foundDecl->getDeclContext()->getSelfProtocolDecl();
+          auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext()->getAsDecl());
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Further nitpick: we've got it set up so you can cast from DeclContext directly to Decl subclasses, so the `getAsDecl()` isn't needed.

Comment (theblixguy) [reply lib/Sema/CSApply.cpp]: Oh! Fixed now.

File: include/swift/AST/DiagnosticsSema.def [hunk 4]
```diff
@@ -550,6 +550,9 @@ ERROR(expr_selector_property_not_settable,none,
       (DescriptiveDeclKind, DeclName))
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
+ERROR(expr_selector_cannot_be_used,none,
+     "cannot use %0 as a selector due to associatedtype requirement on %1",
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Yeah, it might be tricky to check all those cases though. At the moment, it only checks for presence of `associatedtype` members, hence why the diagnostic mentions it. I guess I could reword it to something more specific, like `cannot use %0 as a selector because %1 is not exposed to Objective-C due to associatedtype requirements`?

File: test/FixCode/fixits-apply-objc.swift.result [hunk 5]
```diff
@@ -20,3 +20,18 @@ func foo1(_ an : Any) {
 func foo2(_ messageData: Any?) -> AnyObject? {
   return messageData as AnyObject?
 }
+
+@objc protocol SomeProtocol {
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply test/FixCode/fixits-apply-objc.swift.result]: So you mean we shouldn't be offering a fix-it to the user i.e just a diagnostic error which mentions adding `@objc` to the protocol, but not offering any fix-it?

File: include/swift/AST/DiagnosticsSema.def (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: That's not correct, though. Most protocols aren't exposed to Objective-C even when they can be.

File: test/FixCode/fixits-apply-objc.swift.result (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply test/FixCode/fixits-apply-objc.swift.result]: Oh, sorry. I thought this was a test of what fix-its get auto-applied by the migrator, and it isn't.

File: include/swift/AST/DiagnosticsSema.def (hunk 4 above)
[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Oh, I didn't know that! How about `cannot use %0 as a selector because it is not exposed to Objective-C due to associatedtype requirement on %1`? It's what `expr_selector_not_objc` says and I have just altered it a bit to include `associatedtype`.

[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Hmm, if I don't add this check, then the fixit would be applied on the protocol & then you'll get another error: `Associated type 'Foo' cannot be declared inside '@objc' protocol 'Baz'`. 

I could reword it to something like `cannot use %0 as a selector because %1 is not exposed to Objective-C` & then add a note `It is not possible to expose a protocol to Objective-C if it has associatedtype requirements, non-bridged Swift types, etc`.  

Is this better? or do you have any refinements in mind that I could make to the diagnostic?

[the remark this answers is not in the document]
Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: Honestly, at this point I think it'd be better not to offer the fix-it at all. "cannot use %0 as a selector because protocol %1 is not exposed to Objective-C" is probably the best we can do. I already felt weird about offering a fix-it that might not actually work *and* that might not be what the user wants (such as in the original bug report, where the default implementation won't work because it can't be exposed to Objective-C).

[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: @jrose-apple Hmm seems like a reasonable approach. I have updated the code now!

File: include/swift/AST/DiagnosticsSema.def [hunk 6]
```diff
@@ -551,7 +551,7 @@ ERROR(expr_selector_property_not_settable,none,
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
 ERROR(expr_selector_cannot_be_used,none,
-     "cannot use %0 as a selector due to associatedtype requirement on %1",
+     "cannot use %0 as a selector because protocol %1 is not exposed to Objective C",
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: Nitpick: "Objective-C", with the dash!

File: test/decl/protocol/protocols.swift [hunk 7]
```diff
@@ -513,16 +513,15 @@ extension LetThereBeCrash {
...
+	func anotherFunction()
 }
 
 extension SomeProtocol {
 	
- func someFunction() {
-  let _ = #selector(anotherFunction) // expected-error {{cannot use 'anotherFunction' as a selector due to associatedtype requirement on 'SomeProtocol'}}
- }
+	func someFunction() {
+		let _ = #selector(anotherFunction) // expected-error {{cannot use 'anotherFunction' as a selector because protocol 'SomeProtocol' is not exposed to Objective C}}
```
Comment (jrose-apple) [inline test/decl/protocol/protocols.swift]: You can test that there are no fix-its attached to an `expected-*` diagnostic by adding `{{none}}` after the first `{{...}}`.

(Also nitpick: tabs snuck in here.)

**Review by jrose-apple — COMMENTED** [review]
Thanks for sticking with this one.

File: include/swift/AST/DiagnosticsSema.def (hunk 6 above)
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Fixed!

File: test/decl/protocol/protocols.swift (hunk 7 above)
Comment (theblixguy) [reply test/decl/protocol/protocols.swift]: Fixed!

**Comment by theblixguy** [comment]
@jrose-apple can you invoke the CI?

File: lib/Sema/CSApply.cpp [hunk 8]
```diff
@@ -4196,13 +4196,24 @@ namespace {
...
-                       "@objc ");
+         // If the method declaration lies in a protocol and we're providing
+         // a default implementation of the method through a protocol extension
+         // and using it as a selector, then bail out as adding @objc to the
+         // protocol might not be the right thing to do and could lead to
+         // problems.
+         if (auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext())) {
+             tc.diagnose(E->getLoc(), diag::expr_selector_cannot_be_used,
+                         foundDecl->getBaseName(), protocolDecl->getFullName());
+             return E;
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Nitpick: extra indentation

File: lib/Sema/CSApply.cpp [hunk 9]
```diff
@@ -4196,13 +4196,24 @@ namespace {
...
+         }
+          
+         tc.diagnose(E->getLoc(), diag::expr_selector_not_objc,
+                     foundDecl->getDescriptiveKind(), foundDecl->getFullName())
+            .highlight(subExpr->getSourceRange());
+         tc.diagnose(foundDecl, diag::make_decl_objc,
+                     foundDecl->getDescriptiveKind())
+            .fixItInsert(foundDecl->getAttributeInsertionLoc(false),
+                         "@objc ");
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: Wrong indentation here too. It's either 4-space (because it's a wrapped statement) or 2-space to align the dots.

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by theblixguy** [comment]
@jrose-apple I have fixed the indentation. The tests were failing due because of missing ObjectiveC import, so I have added that to the test file. Can you invoke the CI?

**Comment by harlanhaskins** [comment]
@swift-ci please test

**Comment by theblixguy** [comment]
Seems like adding the ObjectiveC import to that file was a bad idea. I have now moved my test to `expr/primary/selector/selector.swift`, which looks like the right place to put the tests and already contains other related `#selector` tests, so it should all be fine now 🙏  

@harlanhaskins can you invoke the CI again? 😄

**Comment by harlanhaskins** [comment]
@theblixguy 👍

@swift-ci please test

**Comment by theblixguy** [comment]
@harlanhaskins thanks, all tests have passed 🎉

File: lib/Sema/CSApply.cpp [hunk 10]
```diff
@@ -4196,6 +4196,17 @@ namespace {
       // The declaration we found must be exposed to Objective-C.
       tc.validateDecl(method);
       if (!method->isObjC()) {
+        // If the method declaration lies in a protocol and we're providing
+        // a default implementation of the method through a protocol extension
+        // and using it as a selector, then bail out as adding @objc to the
+        // protocol might not be the right thing to do and could lead to
+        // problems.
+        if (auto protocolDecl = dyn_cast<ProtocolDecl>(foundDecl->getDeclContext())) {
+            tc.diagnose(E->getLoc(), diag::expr_selector_cannot_be_used,
```
Comment (jrose-apple) [inline lib/Sema/CSApply.cpp]: I'm sorry, the indentation here is *still* wrong. Two spaces for nested blocks!

File: include/swift/AST/DiagnosticsSema.def [hunk 11]
```diff
@@ -550,6 +550,9 @@ ERROR(expr_selector_property_not_settable,none,
       (DescriptiveDeclKind, DeclName))
 ERROR(expr_selector_property_setter_inaccessible,none,
       "setter of %0 %1 is inaccessible", (DescriptiveDeclKind, DeclName))
+ERROR(expr_selector_cannot_be_used,none,
+     "cannot use %0 as a selector because protocol %1 is not exposed to Objective-C",
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply include/swift/AST/DiagnosticsSema.def]: Done 👍

**Comment by CodaFi** [comment]
LGTM.  Will manually squash this one 

@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@theblixguy Thanks for your patience.  Please update the JIRA ticket.
